### DISCUSSION TITLE
Question about creating dynamic reusable components

### DISCUSSION THREAD

**Discussion by greenshield** [heroui-inc/heroui#1379] [Help] [answered]
I have a question about creating dynamic reusable components.

I have a pretty good grasp on NextUI (I'm building with v2). I have it integrated into my Next.js app as a provider and it works great. I use many of the features, including variants, slots, and have a customized theme.

However, I'm trying to understand the best way to approach something. Here's my use case:

I want to create a customized and reusable dynamic input component.

Obviously, the most basic use of the standard Input componet in NextUI is this:

```
import { Input } from "@nextui-org/react";

<Input label="Email" />
```

I know I can customize the slots like this:

```
import { Input } from "@nextui-org/react";

const classNames = {
    inputWrapper: ["border-2 border-green-500 dark:border-red-500"],
    input: ["text-green-500 dark:text-red-500"],
};

<Input label="Email" classNames={classNames} />
```

And the text and border will be green in light mode and red in dark mode. So far so good.

I want another field with the same style I can reuse the classNames like this:

```
import { Input } from "@nextui-org/react";

const classNames = {
    inputWrapper: ["border-2 border-green-500 dark:border-red-500"],
    input: ["text-green-500 dark:text-red-500"],
};

<Input label="Email" classNames={classNames} />
<Input label="First Name" classNames={classNames} />
```

But if I want to use those classNames on a different Input in a different component I need a way to share them.

A first solution is to wrap the Input in a custom component and spread the props to the Input like this:

```
//CustomInput.jsx
import { Input } from "@nextui-org/react";

const classNames = {
    inputWrapper: ["border-2 border-green-500 dark:border-red-500"],
    input: ["text-green-500 dark:text-red-500"],
};

export default function CustomInput() {
    return <Input classNames={classNames} {...props}>
}

//Profile.jsx
import Input from "./CustomInput";

export default Profile() {
    return <form>
    <Input label ="Username">
    </form>
}

//App.js
import Input from "./CustomInput";
import Profile from "./Profile";

export default App() {
    return <form>
    <Input label ="Email">
    <Input label ="Password">
    </form>

    <Profile />
}
```

This works great, both the Email and Password fields in the App component and the Username field on the imported Profile are all styled correctly, but now comes the question.

I want to use the component, but pass in additional classes. In this case, I simply want to override the green border and make it blue on the Username field. I want to keep the other styling from CustomInput (the green text), but override the border.

The first problem is that if I do something like this:

```
import Input from "./CustomInput";

export default Profile() {
    return <form>
    <Input
          label="Username"
          classNames={{
            inputWrapper: ["border-blue-500"],
          }}
        />
    </form>
}
```

The classNames prop completely overrides the classNames I set in CustomInput, so the border becomes blue the blue text styling is lost.

I thought maybe there was something I could do with extendVariants, but I wasn't able to see a way forward with that. In fact, when I tried to start experimenting the extendVariants import from @nextui-org/ract doesn't exist. Either way, I didn't see how I could make that work for my needs.

I know I can do this in my index.css:

```
@layer base {
  .custom-input-wrapper {
    @apply border-2 border-green-500 dark:border-red-500;
  }

  .custom-input {
    @apply text-green-500 dark:text-red-500;
  }
}
```

Then I can do this is my CustomInput component:

```
//CustomInput.jsx
import { Input } from "@nextui-org/react";

const classNames = {
    inputWrapper: ["custom-input-wrapper"],
    input: ["custom-input"],
};

export default function CustomInput() {
    return <Input classNames={classNames} {...props}>
}

```

But I have the same problem that passing in a new classNames prop will override the ones from the @apply.

So then I thought maybe I could skip creating a CustomInput altogether and somehow define a set of slots as an @apply, something like this, but I don't know if it is even possible to do something like this:

```
@layer base {
  .custom-input-field: {
    inputWrapper: [
        "border-2",
        "border-green-500",
        "dark:border-red-500"
    ],
    input: [
        "text-green-500",
        "dark:text-red-500"
    ]
  }
}
```

And then something like this:

```
<Input label="Username" className="custom-input-field" />
```

So, I'm just wondering what the correct way to approach making a dynamic component that is reusable where I can override the default styles.

Would it be best to detect an incoming className(s) prop in the CustomInput.jsx component and merge them with the defaults?

Which leads to another question. What if I wanted to use the CustomInput component (or the default Input component from NextUI for that matter) and just remove one of the existing default classes. For example, If I wanted <Input label="Username" /> but just remove the border-green-500 class from the inputWrapper? Or if I imported the standard Input component from @nextui-org/react how could I keep everything about the component but remove the "transition-background" and "items-start" classes that are on the inputWrapper for the Input by default?

Anyway, thanks for the help! I'm really happy with NextUI!

**Comment by jrgarciadev** [comment, marked as answer]
Hey @greenshield I'm glad you're enjoying NextUI. Thanks for the detailed explanation,  you have multiple ways to achieve this, but these ones are my preferred options when it comes to creating custom components.


### 1. Destructuring the `classNames` prop

```tsx
import {forwardRef} from 'react';
import {Input, InputProps} from "@nextui-org/react";

const classNames = {
    inputWrapper: ["border-2 border-green-500 dark:border-red-500"],
    input: ["text-green-500 dark:text-red-500"],
};

export interface MyInputProps extends InputProps {}

export const MyInput = forwardRef<HTMLInputElement, MyInputProps>((props, ref) => {
    const {classNames: customClassNames, ...rest} = props;

    const {inputWrapper, input, otherClassnames} = customClassNames || {};

    return <Input ref={ref} classNames={{
        inputWrapper: [...classNames.inputWrapper, ...inputWrapper],
        input: [...classNames.input, ...input],
        ...otherClassnames
    }} {...rest} />;
});

MyInput.displayName = "MyInput";
```



### 2. Using `deepMerge` library

```shell
npm i deepmerge
```

```tsx
import {forwardRef} from 'react';
import {Input, InputProps} from "@nextui-org/react";
import deepMerge from "deepmerge";

const classNames = {
    inputWrapper: ["border-2 border-green-500 dark:border-red-500"],
    input: ["text-green-500 dark:text-red-500"],
};

export interface MyInputProps extends InputProps {}

export const MyInput = forwardRef<HTMLInputElement, MyInputProps>((props, ref) => {
    const {classNames: customClassNames, ...rest} = props;

    const {inputWrapper, input, otherClassnames} = classNames || {};

    return <Input ref={ref} classNames={deepMerge(classNames, customClassNames)} {...rest} />;
});

MyInput.displayName = "MyInput";
```


> **NOTE**: Is important to forward the ref to the original NextUI component

**Reply by greenshield** [reply]
Thanks for the input, I'll try these out!

**Comment by roger-tbg** [comment]
@jrgarciadev - any chance you have a typescript friendly version of the above?
![CleanShot 2024-06-08 at 09 20 12](https://github.com/nextui-org/nextui/assets/141795220/403c13f0-66a4-48c9-ad41-4f83403ef568)
![CleanShot 2024-06-08 at 09 20 05](https://github.com/nextui-org/nextui/assets/141795220/e41bc0cb-e5df-4b25-91ef-3dc4f454f65e)
